src/roll.c

```diff
@@ -0,0 +1,143 @@
+  while (i < nx && (isReal(VECTOR_ELT(x, i)) || isInteger(VECTOR_ELT(x, i)))) i++;
+  if (i != nx)
+    error("x must be list, data.frame or data.table of numeric types");
+
+  nk = length(k);
+  if (nk == 0)                                                            // check that window is not empty
+    error("n must be non 0 length");
+
+  if (!isLogical(adaptive) || length(adaptive) != 1 || LOGICAL(adaptive)[0] == NA_LOGICAL)
+    error("adaptive must be logical TRUE or FALSE");
+
+  if (!LOGICAL(adaptive)[0]) {                                            // validating n input for adaptive=FALSE
+    if (isNewList(k))
+      error("n must be integer, list is accepted for adaptive TRUE");
+    //TODO move R's n=as.integer(n) to C
+    if (!isInteger(k))                                                    // check that k is integer vector
+      error("n must be integer");
+    i = 0;                                                                // check that all window values non-negative
+    while (i < nk && INTEGER(k)[i] >= 0) i++;
+    if (i != nk)
+      error("n must be non-negative integer values (>= 0)");
+  } else if (LOGICAL(adaptive)[0]) {                                      // validating n input for adaptive=TRUE
+    if (isVectorAtomic(k)) {                                              // if not-list then wrap into list
+      kl = allocVector(VECSXP, 1);
+      SET_VECTOR_ELT(kl, 0, k);
+      nk = 1;
+    } else {
+      kl = k;
+    }
+    //TODO move R's n=as.integer(n) to C
+    i = 0;                                                                // check that every column is integer type
+    while (i < nk && isInteger(VECTOR_ELT(kl, i))) i++;
+    if (i != nk)
+      error("n must be list of integer vectors when adaptive TRUE");
+  }
+  
+  if (length(fill) != 1)
+    error("fill must be a vector of length 1");
+
+  if (!strcmp(CHAR(STRING_ELT(align, 0)), "right"))
+    salign = RIGHT;
+  else if (!strcmp(CHAR(STRING_ELT(align, 0)), "center"))
+    salign = CENTER;
+  else if (!strcmp(CHAR(STRING_ELT(align, 0)), "left"))
+    salign = LEFT;
+  else
+    error("Internal error: invalid align argument in rolling function, should have been caught before. please report to data.table issue tracker.");
+
+  if (!isLogical(narm) || length(narm)!=1 || LOGICAL(narm)[0]==NA_LOGICAL)
+    error("na.rm must be TRUE or FALSE");
+  
+  ans = PROTECT(allocVector(VECSXP, nk * nx)); protecti++;                // allocate list to keep results
+  
+  thisfill = PROTECT(coerceVector(fill, REALSXP)); protecti++;
+  
+  if (!LOGICAL(adaptive)[0]) {                                          // adaptive=FALSE
+    if (salign == RIGHT) {                                              // align right scenario
+      for (i=0; i<nx; i++) {                                            // loop over columns
+        this = AS_NUMERIC(VECTOR_ELT(x, i));                            // extract column/vector from data.table/list
```

Comment:
I haven't seen `AS_NUMERIC` used in other places in data.table? Is there any better way to handle integers here? Most commonly double will be as input, so if `AS_NUMERIC` on double is expensive then better to make check and apply only for integers (if any).

Reply:
Correct: I don't recall ever using AS_NUMERIC. If a coerce is needed then `coerceVector()` IIRC and then it's clear a copy of a whole vector is being done.  It's strange to see the AS_NUMERIC()/coerceVector() inside a loop like that in the snippet above.  And VECTOR_ELT() is used for `list()` not vectors,  so this is coercing each item of a list vector it seems.  Oh, this is a loop through each column I guess then,  in which case it makes sense.  Anyway,  that's all that sprang to mind on quick glance.  It's great to be experimenting like this at C level.

Reply:
@mattdowle What if just one column from DT has to be converted
```
SET_VECTOR_ELT(dt, col-1, coerceVector(VECTOR_ELT(dt, col-1)))
```
are this going to update in place?

Reply:
`coerceVector` allocates a new column.  Then `SET_VECTOR_ELT` assigns item `col-1` of the VECSXP dt to point to the new column.  If there is a type change of this column, then there is no safe way (no way in the R API) to avoid allocating the new column.  If the type size was the same (so, not `int` (4) to `double` (8)), then you could in theory change the type of the vector by reference and then manipulate the bits in this vector. But not really worth it.  The coercion is a contiguous scan so I wouldn't worry about churning through one columns worth of RAM for coercion (which is fairly rare anyway).

Reply:
@mattdowle there is type size change, from int to double. But how to change that for just one/few columns without affecting the original data? If I will follow the way it is done in `shallow` I need to create new VECSXP fill it will pointers to doubles from old DT, and convert integer to doubles allocating new columns. Is there better way?

Reply:
Without affecting the original data, yes like in shallow. I'm not aware of a better way, other than coercing int to double on the fly when you need to use those values in that column.